src/commands/BootstrapCommand.js

```diff
@@ -133,32 +134,251 @@ export default class BootstrapCommand extends Command {
+        .map((name) => findPackage(name, pkg.allDependencies[name]) || { name, version: pkg.allDependencies[name] })
+
+        // match external and version mismatched local packages
+        .filter((dep) => !hasPackage(dep.name, dep.version) || !pkg.hasMatchingDependency(dep))
+
+        .forEach(({name, version}) => {
+
+          // Get the object for this package, auto-vivifying.
+          const dep = depsToInstall[name] || (depsToInstall[name] = {
+            versions   : {},
+            dependents : {}
+          });
+
+          // Add this version if it's the first time we've seen it.
+          if (!dep.versions[version]) {
+            dep.versions  [version] = 0;
+            dep.dependents[version] = [];
+          }
+
+          // Record the dependency on this version.
+          dep.versions  [version]++;
+          dep.dependents[version].push(pkg.name);
+        });
+    });
+
+    // determine where each dependency will be installed
+    Object.keys(depsToInstall).forEach((name) => {
+      const {versions, dependents} = depsToInstall[name];
+
+      let rootVersion;
+
+      if (scope && PackageUtilities.getFilteredPackage({name}, {scope, ignore})) {
+
+        // Get the most common version.
+        const commonVersion = Object.keys(versions)
+          .reduce((a, b) => versions[a] > versions[b] ? a : b);
+
+        // Get the version required by the repo root (if any).
+        // If the root doesn't have a dependency on this package then we'll
+        // install the most common dependency there.
+        rootVersion = this.repository.package.allDependencies[name] || commonVersion;
+
+        if (rootVersion !== commonVersion) {
+          this.logger.warn(
+            `The repository root depends on ${name}@${rootVersion}, ` +
+            `which differs from the more common ${name}@${commonVersion}.`
+          );
+        }
+
+        // Install the best version we can in the repo root.
+        // Even if it's already installed there we still need to make sure any
+        // binaries are linked to the packages that depend on them.
+        root.push({
+          name,
+          dependents: (dependents[rootVersion] || [])
+            .map((dep) => this.packageGraph.get(dep).package),
+          dependency: this.repository.hasDependencyInstalled(name, rootVersion)
+            ? null // Don't re-install if it's already there.
+            : `${name}@${rootVersion}`,
+        });
+      }
+
+      // Add less common versions to package installs.
+      Object.keys(versions).forEach((version) => {
+
+        // Only install deps that can't be hoisted in the leaves.
+        if (version === rootVersion) return;
+
+        dependents[version].forEach((pkg) => {
+
+          if (rootVersion) {
+            this.logger.warn(
+              `"${pkg}" package depends on ${name}@${version}, ` +
+              `which differs from the hoisted ${name}@${rootVersion}.`
```

Comment:
yes, this warning has helped me out in the past too, especially if you want to have the same dependency version across all packages.